layers/tls_test.go

```diff
@@ -0,0 +1,243 @@
+	0x69, 0x74, 0x03, 0x93, 0x32, 0xca, 0x54, 0x5e, 0xa5, 0x81, 0x99, 0x6a, 0x73, 0x66, 0xbf, 0x06,
+	0xa0, 0xdc, 0x6a, 0x9c, 0xb1, 0x17, 0x03, 0x01, 0x00, 0x20, 0x44, 0x64, 0xc8, 0xc2, 0x5a, 0xfc,
+	0x4a, 0x82, 0xdd, 0x53, 0x6d, 0x30, 0x82, 0x4d, 0x35, 0x22, 0xf1, 0x5f, 0x3b, 0x96, 0x66, 0x79,
+	0x61, 0x9f, 0x51, 0x93, 0x1b, 0xbf, 0x53, 0x3b, 0xf8, 0x26,
+}
+
+var testAlertEncrypted = []byte{
+	0x15, 0x03, 0x03, 0x00, 0x20, 0x44, 0xb9, 0x9c, 0x2c, 0x6e, 0xab, 0xa3, 0xdf, 0xb1, 0x77, 0x04,
+	0xa2, 0xa4, 0x3a, 0x9a, 0x08, 0x1d, 0xe6, 0x51, 0xac, 0xa0, 0x5f, 0xab, 0x74, 0xa7, 0x96, 0x24,
+	0xfe, 0x62, 0xfe, 0xe8, 0x5e,
+}
+
+// Malformed TLS records
+var testMalformed = []byte{
+	0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f,
+	0x51, 0x52, 0x53, 0x54, 0x55, 0x56, 0x57, 0x58, 0x59, 0x5a, 0x5b, 0x5c, 0x5d, 0x5e, 0x5f,
+	0xd1, 0xd2, 0xd3, 0xd4, 0xd5, 0xd6, 0xd7, 0xd8, 0xd9, 0xda, 0xdb, 0xdc, 0xdd, 0xde, 0xdf,
+}
+
+func TestPacketTLS(t *testing.T) {
+	p := gopacket.NewPacket(testClientHello, LinkTypeEthernet, testDecodeOptions)
+	if p.ErrorLayer() != nil {
+		t.Error("Failed to decode packet:", p.ErrorLayer().Error())
+	}
+
+	checkLayers(p, []gopacket.LayerType{LayerTypeEthernet, LayerTypeIPv4, LayerTypeTCP, LayerTypeTLS}, t)
+}
+
+func TestParseTLSContentType(t *testing.T) {
+	p := gopacket.NewPacket(testClientHello, LinkTypeEthernet, testDecodeOptions)
+	if p.ErrorLayer() != nil {
+		t.Error("Failed to decode packet:", p.ErrorLayer().Error())
+	}
+	checkLayers(p, []gopacket.LayerType{LayerTypeEthernet, LayerTypeIPv4, LayerTypeTCP, LayerTypeTLS}, t)
+
+	r := p.Layer(LayerTypeTLS).(*TLS).Handshake
+	if len(r) != 1 {
+		t.Errorf("Wrong number of handshake records, expected 1, got %d", len(r))
+	}
+
+	ct := r[0].ContentType
+	if ct != TLShandshake {
+		t.Errorf("Failed to parse Content Type, expected %q, got %q", TLShandshake.String(), ct.String())
+	}
+}
+
+func TestParseTLSChangeCipherSpec(t *testing.T) {
+	p := gopacket.NewPacket(testClientKeyExchange, LayerTypeTLS, testDecodeOptions)
+	if p.ErrorLayer() != nil {
+		t.Error("Failed to decode packet:", p.ErrorLayer().Error())
+	}
+	checkLayers(p, []gopacket.LayerType{LayerTypeTLS}, t)
+
+	r := p.Layer(LayerTypeTLS).(*TLS).ChangeCipherSpec
+	if len(r) != 1 {
+		t.Errorf("Wrong number of change cipher spec records, expected 1, got %d", len(r))
+	}
+
+	m := r[0].Message
+	if m == TLSchangecipherspec_unknown {
+		t.Error("Change Cipher Spec message unknown")
+	}
+}
+
+func TestParseTLSAppData(t *testing.T) {
+	p := gopacket.NewPacket(testDoubleAppData, LayerTypeTLS, testDecodeOptions)
+	if p.ErrorLayer() != nil {
+		t.Error("Failed to decode packet:", p.ErrorLayer().Error())
+	}
+	checkLayers(p, []gopacket.LayerType{LayerTypeTLS}, t)
+
+	r := p.Layer(LayerTypeTLS).(*TLS).Appdata
+	if len(r) != 2 {
+		t.Errorf("Wrong number of application data records, expected 2, got %d", len(r))
+	}
+
+	d1 := r[0].Payload
```

Comment:
consider doing a full DeepEqual here (search for DeepEqual in some other tests for an example)